docs/design.md

```diff
@@ -0,0 +1,423 @@
+random-access capability is obtained from the index repo.
+
+In the future, this subcomponent will also serve the top-level cross-shard
+index.
+
+### Persistence and reconciliation
+
+Indices will be persisted in a filesystem directory, referenced by their shard
+key. This makes for easy identification.
+
+Although not expected, it is possible that over time the index repo and the shard management layer state will diverge (entropy):
+
+1. Active shards may have their indices deleted (e.g. if an operator deleted a
+   file or the directory).
+2. Destroyed shards may have their indices lying around (e.g. if there was some
+   system error in housekeeping).
+
+A method `ReconcileIndices()` can be used to reconcile the index repo with the
+shard management layer, by:
+
+1. Fetching gone indices from their mounts.
+2. Deleting lingering indices.
+
+### Interface
+
+```go
+type IndexRepo interface {
+   FullIndexRepo
+   ManifestRepo
+}
+
+type FullIndexRepo interface {
+   // public
+   GetFullIndex(key shard.Key) (FullIndex, error)
+
+   // private, called only by shard management when registering
+   // and destroying shards
+   InsertFullIndex(key shard.Key, index FullIndex) error
+   DeleteFullIndex(key shard.Key) (bool, error)
+}
+
+type ManifestRepo interface {
+   // public
+   GetManifest(key ManifestKey) (Manifest, error)
+   InsertManifest(key ManifestKey, manifest Manifest) error
+   DeleteManifest(key ManifestKey) (bool, error)
+   ExistsManifest(key ManifestKey) (bool, error)
+}
+
+type ManifestKey struct {
+   Shard    key.Shard
+   Rule     string
+   Version  string
+}
+
+type FullIndex interface {
+   Offset(c cid.Cid) (offset int64, err error) // inexistent: offset=-1, err=nil
+   Contains(c cid.Cid) (bool, error)
+   Len() (l int64, err error)
+   ForEach(func(c cid.Cid, offset int64) (ok bool, err error)) error
+}
+```
+
+## DAG access layer
+
+This layer is responsible for serving DAGs or sub-DAGs from one or many shards.
+Initially, queries will require the shard key. That is, queries will be bounded
+to a single identifiable shard.
+
+In the future, when the cross-shard top-level index is implemented, the DAG
+store will be capable of resolving the shard for any given root CID.
+
+DAG access layer allows various access abstractions:
+   1. Obtaining a Blockstore bound to a single shard.
+   2. Obtaining a Blockstore bound to a specific set of shards.
+   3. Obtaining a global Blockstore.
+   4. `ipld.Node` -- TBD.
+
+At this point, we are only concerned with (1). The remaining access patterns
+will be elaborated on in the future.
+
+```go
+type DAGAccessor interface {
+   Shard(key shard.Key) ShardAccessor
+}
+
+type ShardAccessor interface {
+   // Close() the accessor, signalling to the DAG store that we're done here and
+   // possible transient copies can be reclaimed once the shard is released.
+   io.Closer
+
+   Blockstore() (ReadBlockstore, error)
+}
+```
+
+## Requirements for CARv2
+
+- Index needs to be detachable.
```

Comment:
Index readable as a ReaderAt via [`Reader`](https://github.com/ipld/go-car/blob/wip/v2/v2/reader.go#L61).

Do we need similar functionality in the CAR v2 blockstore API?

Reply:
> Do we need similar functionality in the CAR v2 blockstore API?

Mind elaborating what this would look like?

Reply:
A CAR v2 blockstore can be made configurable to either include the index in the final CAR v2 or not, and/or provide the dump of it as an `io.Reader`.

Reply:
Gotcha, you mean on the write-side of the Blockstore. It wouldn't hurt and we might end up using it.

Reply:
yes sorry the write-side of the blockstore I should have been clearer 😊 Thanks @raulk

Reply:
Another related question: would it be OK for the write-side of blockstore to ask for the full list of roots when blockstore is constructed, like [carbon does](https://github.com/willscott/carbon/blob/main/carbon.go#L33)?

Reply:
@masih Yes, we can give you the roots up-front.